### PR TITLE
Use event driven framework

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: src/observers/DocsLoggerObserver.ts [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import { EventType } from "../common/EventType";
+
+export class DocsLoggerObserver {
+    private downloadProgressDot: number;
+    constructor(private logger: OutputChannel) { }
+
+    public eventHandler = (event: BaseEvent) => {
+        switch (event.type) {
+            // Runtime Dependency
+            case EventType.DependencyInstallStart:
```
Comment (yufeih) [inline src/observers/DocsLoggerObserver.ts]: What is the benefit of using this approach? It appears to add a lot of duplicated code and make thing more complex. #Resolved

Comment (928PJY) [reply src/observers/DocsLoggerObserver.ts]: 1. Decouple different component
2. Make input/output of component more clear so it is easier to test

---
In reply to: [362360099](https://github.com/docascode/vscode-docs-build/pull/9#discussion_r362360099) [](ancestors = 362360099)

**Review by yufeih — APPROVED** [review]

File: src/common/EnvironmentController.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import { EventStream } from './EventStream';
+
+const ENVIRONMENT_CONFIG_NAME = 'environment';
+
+export class EnvironmentController implements vscode.Disposable {
+    private environment: Environment;
+    private configurationChangeListener: vscode.Disposable;
+
+    constructor(private eventStream: EventStream) {
+        this.environment = this.getEnv();
```
Comment (live1206) [inline src/common/EnvironmentController.ts]: is `this.` nececssary, can the private fields just start with `_`? #Resolved

File: src/common/EventType.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+export enum EventType {
+    // Runtime dependency
+    DependencyInstallStart,
+    DependencyInstallFinished,
+    PackageInstallStart,
+    PackageInstallSucceeded,
+    PackageInstallFailed,
+    DownloadStart,
```
Comment (live1206) [inline src/common/EventType.ts]: DownloadStarted, same for the others #Resolved

File: src/common/EventType.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+export enum EventType {
+    // Runtime dependency
+    DependencyInstallStart,
+    DependencyInstallFinished,
+    PackageInstallStart,
+    PackageInstallSucceeded,
+    PackageInstallFailed,
+    DownloadStart,
+    DownloadProgress,
```
Comment (live1206) [inline src/common/EventType.ts:9]: DownloadProgressing #Resolved

File: src/common/loggingEvents.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    constructor(public env: Environment) { }
+}
+
+// Runtime Dependency
+export class DependencyInstallStart implements BaseEvent {
+    type = EventType.DependencyInstallStart;
+}
+
+export class DependencyInstallFinished implements BaseEvent {
+    type = EventType.DependencyInstallFinished;
```
Comment (live1206) [inline src/common/loggingEvents.ts:21]: why no constructor for this and the one above? #Resolved

File: src/common/loggingEvents.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import { EventType } from './EventType';
+import { PlatformInformation } from './PlatformInformation';
+import { Environment } from '../shared';
+
+export interface BaseEvent {
```
Comment (live1206) [inline src/common/loggingEvents.ts:5]: can this be a concrete class and only use this one? And the user just set different `EventType` for it #Resolved

File: src/common/EnvironmentController.ts (hunk 2 above)
Comment (928PJY) [reply src/common/EnvironmentController.ts]: it is nessecery.
`_` is not recommended according to this doc: [https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#names](https://github.com/microsoft/TypeScript/wiki/Coding-guidelines#names)

---
In reply to: [362391809](https://github.com/docascode/vscode-docs-build/pull/9#discussion_r362391809) [](ancestors = 362391809)

File: src/common/loggingEvents.ts (hunk 5 above)
Comment (928PJY) [reply src/common/loggingEvents.ts:21]: constructor is not nesscery in ts.

---
In reply to: [362393320](https://github.com/docascode/vscode-docs-build/pull/9#discussion_r362393320) [](ancestors = 362393320)

File: src/common/loggingEvents.ts (hunk 6 above)
Comment (928PJY) [reply src/common/loggingEvents.ts:5]: different event has different property, and `type` is just common property,

---
In reply to: [362393748](https://github.com/docascode/vscode-docs-build/pull/9#discussion_r362393748) [](ancestors = 362393748)

**Review by live1206 — APPROVED** [review]

**Pull request merged**